conductr_cli/sandbox_run_jvm.py

```diff
@@ -148,106 +153,114 @@ def obtain_sandbox_image(image_dir, image_version):
     :param image_version: the version of the sandbox to be downloaded.
     :return: the pair containing path to the expanded core directory and path to the expanded agent directory
     """
-    def resolve_core_binary_from_cache():
+    def resolve_binary_from_cache(binary_cache_path):
         """
-        Checks for the presence of `${cache_dir}/core/${image_version}.tar.gz`.
+        Checks for the presence of the ConductR binary in the cache directory.
 
+        :param binary_cache_path the path of the cached ConductR universal binary.
         :return: If present, return the path to the binary file, else return None.
         """
-        pass
-
-    def resolve_agent_binary_from_cache():
-        """
-        Checks for the presence of `${cache_dir}/agent/${image_version}.tar.gz`.
-
-        :return: If present, return the path to the binary file, else return None.
-        """
-        pass
+        if os.path.exists(binary_cache_path):
+            return binary_cache_path
+        else:
+            return None
 
-    def download_core_from_bintray(bintray_auth):
+    def download_from_bintray(bintray_auth, bintray_package_name):
         """
-        Downloads core from bintray given `${image_version}`.
+        Downloads ConductR binary from Bintray given `${binary_info}`.
 
-        The images are expected to be cached in `${cache_dir}/core`.
+        The images are expected to be cached in `${image_dir}`.
 
-        If the binary is not yet available within the local cache, then it will be downloaded from Bintray.
-
-        The aftefacts will be available under the following Bintray repo:
+        The artifacts are available under the following Bintray repo:
 
         https://bintray.com/lightbend/commercial-releases/ConductR-Universal
 
         As part of the download:
         - A progress bar will be displayed.
-        - The download will be saved into `${cache_dir}/core/${image_version}.tar.gz.tmp`. Once download is complete,
-          this file will be moved to `${cache_dir}/core/${image_version}.tar.gz`.
+        - The download will be saved into `${image_dir}/${filename}.tgz.tmp`.
+        Once download is complete, this file will be moved to `${image_dir}/${filename}.tgz`.
 
         :param bintray_auth: a tuple containing Bintray auth information which can be passed into the request library.
                              This information is loaded from bintray credentials file, which can be None if the
                              credentials is not specified
+        :param bintray_package_name: the Bintray package name of the ConductR binary
         :return: path to the downloaded file
         """
-        pass
+        _, _, cached_file = bintray_resolver.bintray_download(image_dir, BINTRAY_LIGHTBEND_ORG, BINTRAY_CONDUCTR_REPO,
```

Comment:
I think we need a condition which handles a download failure at this point.

When a download is successful, it will return `True, file_name, cached_file` when failed, it will return `False, None, None`.

With the current change, when the download failed, the `cached_file` will be set to `None`, which will cause the npe-like error with the following stack trace:

```
2017-01-11 10:37:46,759: Failure running the following command: ['/usr/local/bin/sandbox', 'run', '2.0.0-rc2']
Traceback (most recent call last):
  File "/Users/felixsatyaputra/workspace/typesafe-fsat/conductr-cli/conductr_cli/main_handler.py", line 14, in run
    result = callback()
  File "/Users/felixsatyaputra/workspace/typesafe-fsat/conductr-cli/conductr_cli/sandbox.py", line 6, in main_method
    sandbox_main.run()
  File "/Users/felixsatyaputra/workspace/typesafe-fsat/conductr-cli/conductr_cli/sandbox_main.py", line 276, in run
    result = args.func(args)
  File "/Users/felixsatyaputra/workspace/typesafe-fsat/conductr-cli/conductr_cli/validation.py", line 36, in handler
    return func(*args, **kwargs)
  File "/Users/felixsatyaputra/workspace/typesafe-fsat/conductr-cli/conductr_cli/validation.py", line 52, in handler
    return func(*args, **kwargs)
  File "/Users/felixsatyaputra/workspace/typesafe-fsat/conductr-cli/conductr_cli/validation.py", line 306, in handler
    return func(*args, **kwargs)
  File "/Users/felixsatyaputra/workspace/typesafe-fsat/conductr-cli/conductr_cli/validation.py", line 325, in handler
    return func(*args, **kwargs)
  File "/Users/felixsatyaputra/workspace/typesafe-fsat/conductr-cli/conductr_cli/sandbox_run.py", line 44, in run
    run_result = sandbox_run_jvm.run(args, features)
  File "/Users/felixsatyaputra/workspace/typesafe-fsat/conductr-cli/conductr_cli/sandbox_run_jvm.py", line 32, in run
    core_extracted_dir, agent_extracted_dir = obtain_sandbox_image(args.image_dir, args.image_version)
  File "/Users/felixsatyaputra/workspace/typesafe-fsat/conductr-cli/conductr_cli/sandbox_run_jvm.py", line 260, in obtain_sandbox_image
    core_extracted_dir = extract_binary(core_binary_path, core_binary_info['extraction_dir'])
  File "/Users/felixsatyaputra/workspace/typesafe-fsat/conductr-cli/conductr_cli/sandbox_run_jvm.py", line 236, in extract_binary
    shutil.unpack_archive(binary_path, extraction_dir)
  File "/usr/local/Cellar/python3/3.5.2/Frameworks/Python.framework/Versions/3.5/lib/python3.5/shutil.py", line 964, in unpack_archive
    format = _find_unpack_format(filename)
  File "/usr/local/Cellar/python3/3.5.2/Frameworks/Python.framework/Versions/3.5/lib/python3.5/shutil.py", line 932, in _find_unpack_format
    if filename.endswith(extension):
AttributeError: 'NoneType' object has no attribute 'endswith'
```